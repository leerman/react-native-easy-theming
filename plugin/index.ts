import { PluginObj } from "@babel/core";
import * as types from "@babel/types";
import { getStyleValue } from "./utils/get-style-value";
import { mapStyleName } from "./utils/map-style-name";
import { transformStyleName } from "./utils/transform-style-name";
import { addThemeImports } from "./utils/add-theme-imports";

const STYLES_PREFIX = "s-";

export default function (babel: { types: typeof types }): PluginObj {
  var t = babel.types;

  return {
    visitor: {
      JSXOpeningElement(path, { opts }) {
        const prefix = (opts as any).prefix ?? STYLES_PREFIX;
        const themeIdentifier = path.scope.generateUidIdentifier("theme");

        if (
          path.node.name.type === "JSXIdentifier" &&
          path.node.name.name === "Text"
        ) {
          let styleAttribute;
          let newStyleObject;
          let shouldAddThemeImport = false;

          path.node.attributes.forEach(function (attr) {
            if (attr.type === "JSXSpreadAttribute") return;

            if (attr.name.name === "style") {
              // console.log('style', JSON.stringify(attr, null, 2));
            }
            if (
              typeof attr.name.name === "string" &&
              attr.name.name.startsWith(prefix)
            ) {
              const rawStyleName = attr.name.name.slice(prefix.length);
              const styleName = transformStyleName(mapStyleName(rawStyleName));

              if (!styleAttribute) {
                styleAttribute = t.jSXAttribute(
                  t.jSXIdentifier("style"),
                  t.jSXExpressionContainer(t.objectExpression([]))
                );

                newStyleObject = styleAttribute.value.expression;
                path.node.attributes.push(styleAttribute);
              }

              const { addThemeImport, value } = getStyleValue({
                attribute: attr,
                themeIdentifier,
                key: rawStyleName,
                t,
              });

              newStyleObject.properties.push(
                t.objectProperty(t.identifier(styleName), value)
              );

              if (addThemeImport) {
                shouldAddThemeImport = true;
              }

              // newStyleObject.properties.push(
              //   t.objectProperty(
              //     t.identifier(styleName),
              //     t.isJSXExpressionContainer(attr.value)
              //       ? attr.value.expression
              //       : isColor((attr.value as any).value)
              //       ? t.stringLiteral((attr.value as any).value)
              //       : (attr.value as any).value
              //           .split(".")
              //           .reduce(
              //             (prev, next) =>
              //               t.memberExpression(prev, t.identifier(next)),
              //             themeIdentifier
              //           )
              //   )
              // );
              path.node.attributes.splice(
                path.node.attributes.indexOf(attr),
                1
              );
            }
          });

          if (shouldAddThemeImport) {
            addThemeImports({ path, t, themeIdentifier });
          }
        }
      },
    },
  };
}
