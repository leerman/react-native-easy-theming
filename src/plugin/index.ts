import { PluginObj, types } from "@babel/core";
import { getStyleValue } from "./utils/get-style-value";
import { mapStyleName } from "./utils/map-style-name";
import { transformStyleName } from "./utils/transform-style-name";
import { addThemeImports } from "./utils/add-theme-imports";

const STYLES_PREFIX = "s-";

export default function (babel: { types: typeof types }): PluginObj {
  var t = babel.types;

  return {
    visitor: {
      JSXOpeningElement(path, { opts }) {
        const prefix = (opts as any).prefix ?? STYLES_PREFIX;
        const themeIdentifier = path.scope.generateUidIdentifier("theme");

        if (path.node.name.type === "JSXIdentifier") {
          let shouldAddThemeImport = false;
          const beforeStyleProps = [];
          const afterStyleProps = [];
          let existsStyleAttribute = null;
          const deleteIndexes = [];

          path.node.attributes.forEach(function (attr, index) {
            if (attr.type === "JSXSpreadAttribute") return;

            if (attr.name.name === "style") {
              if (attr.value.type !== "JSXExpressionContainer") {
                throw new Error("unhandled style prop: " + attr.value.type);
              }
              existsStyleAttribute = attr.value.expression;
              deleteIndexes.push(index);
            }
            if (
              typeof attr.name.name === "string" &&
              attr.name.name.startsWith(prefix)
            ) {
              const rawStyleName = attr.name.name.slice(prefix.length);
              const styleName = transformStyleName(mapStyleName(rawStyleName));

              const { addThemeImport, value } = getStyleValue({
                path,
                attribute: attr,
                themeIdentifier,
                key: rawStyleName,
                t,
              });

              if (existsStyleAttribute !== null) {
                afterStyleProps.push(
                  t.objectProperty(t.identifier(styleName), value)
                );
              } else {
                beforeStyleProps.push(
                  t.objectProperty(t.identifier(styleName), value)
                );
              }

              deleteIndexes.push(index);

              if (addThemeImport) {
                shouldAddThemeImport = true;
              }
            }
          });

          if (existsStyleAttribute !== null) {
            const hasExtra =
              beforeStyleProps.length > 0 || afterStyleProps.length > 0;
            path.node.attributes.push(
              t.jSXAttribute(
                t.jSXIdentifier("style"),
                t.jSXExpressionContainer(
                  !hasExtra
                    ? existsStyleAttribute
                    : t.arrayExpression(
                        [
                          beforeStyleProps.length > 0
                            ? t.objectExpression(beforeStyleProps)
                            : null,
                          existsStyleAttribute,
                          afterStyleProps.length > 0
                            ? t.objectExpression(afterStyleProps)
                            : null,
                        ].filter((item) => item !== null)
                      )
                )
              )
            );
          } else if (beforeStyleProps.length > 0) {
            path.node.attributes.push(
              t.jSXAttribute(
                t.jSXIdentifier("style"),
                t.jSXExpressionContainer(t.objectExpression(beforeStyleProps))
              )
            );
          }

          deleteIndexes.reverse().forEach((index) => {
            path.node.attributes.splice(index, 1);
          });

          if (shouldAddThemeImport) {
            addThemeImports({ path, t, themeIdentifier });
          }
        }
      },
    },
  };
}
