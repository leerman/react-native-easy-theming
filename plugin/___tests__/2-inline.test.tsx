import { pluginTester } from "babel-plugin-tester";
import plugin from "../index";

pluginTester({
  title: "not apply",
  plugin,
  snapshot: true,
  babelOptions: {
    presets: ["@babel/preset-react"],
  },
  tests: [
    {
      title: "without attributes",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          return <Text>hello</Text>
        }
      `,
    },
    {
      title: "with plane color value",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          return <Text s-background-color="red">hello</Text>
        }
      `,
    },
  ],
});
