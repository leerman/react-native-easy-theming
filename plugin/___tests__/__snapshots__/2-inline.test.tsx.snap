// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`not apply 1. without attributes: 1. without attributes 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(Text, null, "hello");
};

`;

exports[`not apply 2. with plane color value: 2. with plane color value 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text s-background-color="red">hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        backgroundColor: "red",
      },
    },
    "hello"
  );
};

`;

exports[`not apply 3. with value inside expression: 3. with value inside expression 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text s-background-color={"red"}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        backgroundColor: "red",
      },
    },
    "hello"
  );
};

`;
